cpp/box/Box.h

```diff
@@ -336,12 +336,17 @@ class Box
     /*! \param vecs Vectors to wrap, updated to the minimum image obeying the periodic settings
      *  \param Nvecs Number of vectors
      */
-    void wrap(vec3<float>* vecs, unsigned int Nvecs) const
+    void wrap(vec3<float>* vecs, unsigned int Nvecs, vec3<float> *out = nullptr) const
     {
+        if (out == nullptr)
```

Comment:
An alternative approach would be for us to change the signature of this method, and _require_ `out` (no default value of `nullptr`). This would allow us to make the first argument `const vec3<float>*` (constant). This method is only used by Cython; I could not find any instances of `wrap` applied to multiple vectors in the C++ code. @vyasr @Charlottez112 What do you think?

Reply:
Yeah I considered that possibility as well, and chose to stick with this implementation precisely because I wanted to maintain `out` as an optional parameter. However, on second though I've changed my mind and would like to require the `out` parameter in C++. Here's my reasoning: we already have different semantics in Python and C++, because in Python we create a new array for the user if one is not provided. I think it would be a less obvious, and therefore more insidious, inconsistency if the default behavior when `out` is not provided is to overwrite in place in C++ while it is to return a copy in Python. C++ users are much more likely to care about the performance impact of minimizing copies, so it's much more consistent with best practices in the language to require them to handle allocating new arrays if they need to; otherwise, they can just pass a pointer to the same array to trigger in-place overwrites.

Note that in order for this to be a good idea we'd want to make sure that all the C++ APIs for box functions behave the same way. I think that shouldn't be an issue though.

Reply:
I agree with @vyasr. I'm not sure I understand the benefits of requiring ```out```, and I don't know enought about the C++ user cases to make comments on that. But one thing I've been thinking is that if we're doing all the checks (in ```box.wrap```) for ```out``` when it's given, why don't we just make a new array for it in ```box.wrap``` instead of making one in ```_convert_array```, when it's the default ```None```.

Reply:
@Charlottez112 I'm not sure I follow, which part of what I said are you agreeing with? It sounds like not the part about requiring `out`, so I'm a bit confused.

Regarding where we make the new array, I assume that you mean in `box.wrap` in Python, not in the C++ function? If so, the main reason I suggested doing it in `_convert_array` is because then the exact same logic can be used for all the other box functions. Anywhere that a function requires an array to have certain properties but where a user might not pass that array, those functions could rely on `_convert_array` to create an appropriate array. Does that make sense?

Reply:
@bdice let me know whether we're on the same page here before we ask @Charlottez112 to make any changes.

Reply:
@vyasr I'm on board with the proposal to require `out` in C++, and particularly this line of reasoning that you mentioned above:

> I think it would be a less obvious, and therefore more insidious, inconsistency if the default behavior when out is not provided is to overwrite in place in C++ while it is to return a copy in Python.

I don't entirely understand the subsequent discussion, it sounds like there is some confusion / other topics. Please let me know if you're asking for input on some piece of that.

Reply:
> @Charlottez112 I'm not sure I follow, which part of what I said are you agreeing with? It sounds like not the part about requiring `out`, so I'm a bit confused.

Sorry for the confusion. I meant to ask @bdice if he could explain a bit more about what he thinks.

> Regarding where we make the new array, I assume that you mean in `box.wrap` in Python, not in the C++ function? If so, the main reason I suggested doing it in `_convert_array` is because then the exact same logic can be used for all the other box functions. Anywhere that a function requires an array to have certain properties but where a user might not pass that array, those functions could rely on `_convert_array` to create an appropriate array. Does that make sense?

Yeah that makes sense. I guess I just don't know if there's a better way to combine checking the ```out``` array with generating one if it's none.

Reply:
I went ahead and changed this part in ```Box.h```. Not sure if there's anything else we need to change